src/core/qgsdiagramrenderer.cpp

```diff
@@ -804,6 +851,275 @@ void QgsLinearlyInterpolatedDiagramRenderer::writeXml( QDomElement &layerElem, Q
+{
+  if ( !mDiagram )
+  {
+    return;
+  }
+
+  QPointF newPos = pos; // Each subdiagram will have its own newPos
+
+  // Get subrenderers sorted by mode (vertical diagrams are returned backwards)
+  const QList< QgsDiagramRenderer * > stackedRenderers = renderers( true );
+
+  for ( const auto &stackedRenderer : stackedRenderers )
+  {
+    if ( stackedRenderer->rendererName() == QStringLiteral( "Stacked" ) )
+    {
+      // Nested stacked diagrams will use this recursion
+      stackedRenderer->renderDiagram( feature, c, newPos, properties );
+      continue;
+    }
+
+    QgsDiagramSettings s;
+    if ( !stackedRenderer->diagramSettings( feature, c, s ) )
+    {
+      continue;
+    }
+
+    if ( !s.enabled )
+    {
+      continue;
+    }
+
+    if ( s.scaleBasedVisibility )
+    {
+      double maxScale = s.maximumScale;
+      if ( maxScale > 0 && c.rendererScale() < maxScale )
+      {
+        continue;
+      }
+      double minScale = s.minimumScale;
+      if ( minScale > 0 && c.rendererScale() > minScale )
+      {
+        continue;
+      }
+    }
+
+    if ( properties.hasActiveProperties() )
+    {
+      c.expressionContext().setOriginalValueVariable( QgsColorUtils::colorToString( s.backgroundColor ) );
+      s.backgroundColor = properties.valueAsColor( QgsDiagramLayerSettings::Property::BackgroundColor, c.expressionContext(), s.backgroundColor );
+      c.expressionContext().setOriginalValueVariable( QgsColorUtils::colorToString( s.penColor ) );
+      s.penColor = properties.valueAsColor( QgsDiagramLayerSettings::Property::StrokeColor, c.expressionContext(), s.penColor );
+      c.expressionContext().setOriginalValueVariable( s.penWidth );
+      s.penWidth = properties.valueAsDouble( QgsDiagramLayerSettings::Property::StrokeWidth, c.expressionContext(), s.penWidth );
+      c.expressionContext().setOriginalValueVariable( s.rotationOffset );
+      s.rotationOffset = properties.valueAsDouble( QgsDiagramLayerSettings::Property::StartAngle, c.expressionContext(), s.rotationOffset );
+    }
+
+    QgsPaintEffect *effect = s.paintEffect();
+    std::unique_ptr< QgsEffectPainter > effectPainter;
+    if ( effect && effect->enabled() )
+    {
+      effectPainter = std::make_unique< QgsEffectPainter >( c, effect );
+    }
+
+    stackedRenderer->diagram()->renderDiagram( feature, c, s, newPos );
+    QgsStackedDiagram *stackedDiagram = dynamic_cast< QgsStackedDiagram *>( mDiagram.get() );
+    stackedDiagram->subDiagramPosition( newPos, c, mSettings, s );
+  }
+}
+
+bool QgsStackedDiagramRenderer::diagramSettings( const QgsFeature &feature, const QgsRenderContext &c, QgsDiagramSettings &s ) const
+{
+  Q_UNUSED( feature )
+  Q_UNUSED( c )
+  Q_UNUSED( s )
+  return false;
+}
+
+QSizeF QgsStackedDiagramRenderer::diagramSize( const QgsFeature &feature, const QgsRenderContext &c ) const
+{
+  Q_UNUSED( feature )
+  Q_UNUSED( c )
+  return QSizeF( 0, 0 );
```

Comment:
Why is this does needed to go through the children and return the global size?

Reply:
`diagramSize` is protected, it'll only be called inside the public method `sizeMapUnits()`, which we do override for Stack Diagrams, delegating to subrenderers and adding up their sizes to get the stacked diagram size.

So, for the Stacked Diagram Renderer, the `diagramSize()` is never called, and that's why we do a dummy implementation returning a default value (it's a pure virtual method, so we needed to override it).

Reply:
In this case, would it make sense to make it `private` instead of `protected`?

Reply:
Yeap, it's existing API. It's in the abstract superclass, it must be overridden by subclasses.